Quarkus NativeImageBuildStep fails with perm denied with docker rootless

### Describe the bug

Building a minimal Quarkus project in native mode via docker rootless fails.

### Expected behavior

The build should not fail.

### Actual behavior

The build runs:
```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] docker run --env LANG=C --rm --user 1000:1000 -v /home/rhousni/projects/quarkus-poc/target/to-csv-1.0-SNAPSHOT-native-image-source-jar:/project:z --name build-native-bENIH quay.io/quarkus/ubi-quarkus-native-image:22.1-java17 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Duser.language=en -J-Duser.country=US -J-Dfile.encoding=UTF-8 -H:-ParseOnce -J--add-exports=java.security.jgss/sun.security.krb5=ALL-UNNAMED -J--add-opens=java.base/java.text=ALL-UNNAMED -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -H:+AllowFoldMethods -J-Djava.awt.headless=true -H:FallbackThreshold=0 --link-at-build-time -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:EnableURLProtocols=http -H:NativeLinkerOption=-no-pie -H:-UseServiceLoaderFeature -H:+StackTrace to-csv-1.0-SNAPSHOT-runner -jar to-csv-1.0-SNAPSHOT-runner.jar
```

then fails with:
```
/gcc/x86_64-redhat-linux/8/crtend.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crtn.o
/usr/bin/ld: cannot open output file /project/to-csv-1.0-SNAPSHOT-runner: Permission denied
collect2: error: ld returned 1 exit status
	at com.oracle.svm.hosted.image.NativeImageViaCC.handleLinkerFailure(NativeImageViaCC.java:505)
	at com.oracle.svm.hosted.image.NativeImageViaCC.write(NativeImageViaCC.java:452)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:695)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:515)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:615)
------------------------------------------------------------------------------------------------------------------------
                        1.5s (4.2% of total time) in 23 GCs | Peak RSS: 6.19GB | CPU load: 10.81
========================================================================================================================
Failed generating 'to-csv-1.0-SNAPSHOT-runner' after 35.7s.
Error: Image build request failed with exit status 1
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  42.163 s
[INFO] Finished at: 2022-07-22T19:55:56+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.10.3.Final:build (default) on project to-csv: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: io.quarkus.deployment.pkg.steps.NativeImageBuildStep$ImageGenerationFailureException: Image generation failed. Exit code: 1
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.imageGenerationFailed(NativeImageBuildStep.java:400)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:240)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:944)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

### How to Reproduce?

1. Make sure `docker` is installed in rootless mode => https://docs.docker.com/engine/security/rootless/
2. Create an empty Quarkus Maven project
3. Run `./mvnw clean package -P native`

### Output of `uname -a` or `ver`

```
Linux rhousni 5.15.0-41-generic #44~20.04.1-Ubuntu SMP Fri Jun 24 13:27:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Output of `java -version`

```
openjdk version "17" 2021-09-14 OpenJDK Runtime Environment (build 17+35-2724) OpenJDK 64-Bit Server VM (build 17+35-2724, mixed mode, sharing)
```

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

`2.10.3`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63) Maven home: /home/rhousni/.m2/wrapper/dists/apache-maven-3.8.6-bin/67568434/apache-maven-3.8.6 Java version: 17.0.1, vendor: Eclipse Adoptium, runtime: /home/rhousni/.jdks/temurin-17.0.1 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.15.0-41-generic", arch: "amd64", family: "unix"
```

### Additional information

Maybe linked to #10637 ?

--- a/core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java
@@ -1,11 +1,8 @@
 
 package io.quarkus.deployment;
 
-import java.io.BufferedReader;
 import java.io.File;
 import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
 import java.lang.reflect.InvocationTargetException;
 import java.net.InetSocketAddress;
 import java.net.Socket;
@@ -15,7 +12,6 @@ import java.time.Duration;
 import java.util.List;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
-import java.util.function.Function;
 import java.util.function.Supplier;
 
 import org.eclipse.microprofile.config.ConfigProvider;
@@ -176,29 +172,6 @@ public class IsDockerWorking implements BooleanSupplier {
             }
         }
 
-        public static class OutputFilter implements Function<InputStream, Runnable> {
-            private final StringBuilder builder = new StringBuilder();
-
-            @Override
-            public Runnable apply(InputStream is) {
-                return () -> {
-
-                    try (InputStreamReader isr = new InputStreamReader(is);
-                            BufferedReader reader = new BufferedReader(isr)) {
-
-                        for (String line = reader.readLine(); line != null; line = reader.readLine()) {
-                            builder.append(line);
-                        }
-                    } catch (IOException e) {
-                        throw new RuntimeException("Error reading stream.", e);
-                    }
-                };
-            }
-
-            public String getOutput() {
-                return builder.toString();
-            }
-        }
     }
 
     private enum Result {

--- a/core/deployment/src/main/java/io/quarkus/deployment/OutputFilter.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/OutputFilter.java
@@ -1,0 +1,31 @@
+package io.quarkus.deployment;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.function.Function;
+
+public class OutputFilter implements Function<InputStream, Runnable> {
+    private final StringBuilder builder = new StringBuilder();
+
+    @Override
+    public Runnable apply(InputStream is) {
+        return () -> {
+
+            try (InputStreamReader isr = new InputStreamReader(is);
+                    BufferedReader reader = new BufferedReader(isr)) {
+
+                for (String line = reader.readLine(); line != null; line = reader.readLine()) {
+                    builder.append(line);
+                }
+            } catch (IOException e) {
+                throw new RuntimeException("Error reading stream.", e);
+            }
+        };
+    }
+
+    public String getOutput() {
+        return builder.toString();
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
@@ -2,35 +2,110 @@ package io.quarkus.deployment.pkg.steps;
 
 import static io.quarkus.deployment.pkg.steps.LinuxIDUtil.getLinuxID;
 
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
 import java.nio.file.Path;
+import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.SystemUtils;
+import org.jboss.logging.Logger;
 
+import io.quarkus.deployment.OutputFilter;
 import io.quarkus.deployment.pkg.NativeConfig;
+import io.quarkus.deployment.util.ExecUtil;
 import io.quarkus.deployment.util.FileUtil;
 import io.quarkus.runtime.util.ContainerRuntimeUtil;
 
 public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContainerRunner {
 
+    private static final Logger LOGGER = Logger.getLogger(NativeImageBuildLocalContainerRunner.class.getName());
+
     public NativeImageBuildLocalContainerRunner(NativeConfig nativeConfig, Path outputDir) {
         super(nativeConfig, outputDir);
         if (SystemUtils.IS_OS_LINUX) {
             ArrayList<String> containerRuntimeArgs = new ArrayList<>(Arrays.asList(baseContainerRuntimeArgs));
-            String uid = getLinuxID("-ur");
-            String gid = getLinuxID("-gr");
-            if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
-                Collections.addAll(containerRuntimeArgs, "--user", uid + ":" + gid);
-                if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
-                    // Needed to avoid AccessDeniedExceptions
-                    containerRuntimeArgs.add("--userns=keep-id");
+            if (isDockerRootless(containerRuntime)) {
+                Collections.addAll(containerRuntimeArgs, "--user", String.valueOf(0));
+            } else {
+                String uid = getLinuxID("-ur");
+                String gid = getLinuxID("-gr");
+                if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
+                    Collections.addAll(containerRuntimeArgs, "--user", uid + ":" + gid);
+                    if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
+                        // Needed to avoid AccessDeniedExceptions
+                        containerRuntimeArgs.add("--userns=keep-id");
+                    }
                 }
             }
             baseContainerRuntimeArgs = containerRuntimeArgs.toArray(baseContainerRuntimeArgs);
         }
+    }
+
+    private static boolean isDockerRootless(ContainerRuntimeUtil.ContainerRuntime containerRuntime) {
+        if (containerRuntime != ContainerRuntimeUtil.ContainerRuntime.DOCKER) {
+            return false;
+        }
+        String dockerEndpoint = fetchDockerEndpoint();
+        // docker socket?
+        String socketUriPrefix = "unix://";
+        if (dockerEndpoint == null || !dockerEndpoint.startsWith(socketUriPrefix)) {
+            return false;
+        }
+        String dockerSocket = dockerEndpoint.substring(socketUriPrefix.length());
+        String currentUid = getLinuxID("-ur");
+        if (currentUid == null || currentUid.isEmpty() || currentUid.equals(String.valueOf(0))) {
+            return false;
+        }
+
+        int socketOwnerUid;
+        try {
+            socketOwnerUid = (int) Files.getAttribute(Path.of(dockerSocket), "unix:uid", LinkOption.NOFOLLOW_LINKS);
+        } catch (IOException e) {
+            LOGGER.infof("Owner UID lookup on '%s' failed with '%s'", dockerSocket, e.getMessage());
+            return false;
+        }
+        return currentUid.equals(String.valueOf(socketOwnerUid));
+    }
+
+    private static String fetchDockerEndpoint() {
+        // DOCKER_HOST environment variable overrides the active context
+        String dockerHost = System.getenv("DOCKER_HOST");
+        if (dockerHost != null) {
+            return dockerHost;
+        }
+
+        OutputFilter outputFilter = new OutputFilter();
+        if (!ExecUtil.execWithTimeout(new File("."), outputFilter, Duration.ofMillis(3000),
+                "docker", "context", "ls", "--format",
+                "'{{- if .Current -}} {{- .DockerEndpoint -}} {{- end -}}'")) {
+            LOGGER.debug("Docker context lookup didn't succeed in time");
+            return null;
+        }
+
+        Set<String> endpoints = outputFilter.getOutput().lines()
+                .filter(Objects::nonNull)
+                .filter(Predicate.not(String::isBlank))
+                .collect(Collectors.toSet());
+        if (endpoints.size() == 1) {
+            return endpoints.stream().findFirst().orElse(null);
+        }
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debugf("Found too many active Docker endpoints: [%s]",
+                    endpoints.stream()
+                            .map(endpoint -> String.format("'%s'", endpoint))
+                            .collect(Collectors.joining(",")));
+        }
+        return null;
     }
 
     @Override
@@ -45,4 +120,5 @@ public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContai
                 volumeOutputPath + ":" + NativeImageBuildStep.CONTAINER_BUILD_VOLUME_PATH + ":z");
         return containerRuntimeArgs;
     }
+
 }

